File: lib/lightdm.dart
import 'dart:js';

import 'package:js/js.dart';
import 'package:lightdm_webkit_theme_litarvan/desktop.dart';
import 'package:lightdm_webkit_theme_litarvan/user.dart';

class LightDM {
  Function errorCallback;
  JsObject lightdm;

  List<Desktop> _desktops;
  List<User> _users;

  LightDM(this.errorCallback) {
    this.lightdm = context['lightdm'];
    context.callMethod('init', [allowInterop(errorCallback)]);
  }

  login(String username, String password) => context.callMethod('login', [username, password]);
  start(String desktop) => context.callMethod('start', [desktop]);
  shutdown() => lightdm.callMethod('shutdown');
  reboot() => lightdm.callMethod('restart');

  List<Desktop> get desktops {
    if (_desktops == null) {
      _desktops = [];

      for (JsObject obj in lightdm['sessions']) {
        _desktops.add(new Desktop(obj['key'], obj['name']));
      }
    }

    return _desktops;
  }

  List<User> get users {
    if (_users == null) {
      _users = [];

      for (JsObject obj in lightdm['users']) {
        _users.add(new User(obj['display_name'], obj['username'], obj['image'] != null ? (obj['image'].startsWith('/') ? 'file://' : '') + obj['image'] : 'images/default_user.png'));
      }
    }

    return _users;
  }
}

File: js/lightdm.js
'use strict';

window.debug = window.lightdm === undefined;

if (window.debug)
{
    window.lightdm = {
        password_debugged: false,
        is_authenticated: false,
        authentication_user: 'DEBUG USER',
        default_session: 'plasma-shell',
        sessions: [
            {
                name: 'KDE 5',
                key: 'plasma-shell'
            },
            {
                name: 'Gnome 3',
                key: 'gnome-shell'
            },
            {
                name: 'XFCE 4',
                key: 'xfce'
            }
        ],
        users: [{
            name: 'litarvan'
        }],
        start_authentication: (username) => {
            console.log(`Starting authenticating : '${username}'`);
            show_prompt("Password:");
        },
        cancel_authentication: () => {
            console.log('Auth cancelled');
        },
        respond: (password) => {
            console.log(`Password provided : '${password}'`);

            if (lightdm.password_debugged)
            {
                lightdm.is_authenticated = true;
            }
            else
            {
                lightdm.password_debugged = true;
            }

            authentication_complete();
        },
        login: (user, session) => {
            alert(`Logged with '${user}' (Session: '${session}') !`);
        },
        shutdown: () => {
            alert('(DEBUG: System is shutting down)')
        },
        restart: () => {
            alert('(DEBUG: System is rebooting)')
        }
    };
}

let password;
let errorCB;
let completeCB;

function login(username, pass, cb)
{
    lightdm.start_authentication(username);
    password = pass;

    completeCB = cb;
}

function start(desktop)
{
    lightdm.login(lightdm.authentication_user, desktop);
}

function show_prompt(text, type)
{
    if (text === "Password:")
    {
        lightdm.respond(password);
    }
}

function authentication_complete()
{
    if (lightdm.is_authenticated)
    {
        lightdm.cancel_authentication();
        errorCB("Invalid username/password");
    }
    else
    {
        completeCB(lightdm.authentication_user);
    }
}

function show_message(text, type)
{
    errorCB(text);
}

window.init = (errorCallback) =>
{
    console.log(errorCallback);
    errorCB = errorCallback;
};